File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseBetween(self, head, left, right):
        """
        :type head: ListNode
        :type left: int
        :type right: int
        :rtype: ListNode
        """
        sanitel = ListNode(-1)
        sanitel.next = head
        prev, cur = None, sanitel
        start, end = None, None
        ind = 0
        left_node = None
        while(ind <= right):
            if ind == left - 1:
                left_node = cur
            if ind == left:
                start = cur
            if ind == right:
                end = cur
            nxt = cur.next
            if ind >= left + 1:
                cur.next = prev
            prev = cur
            cur = nxt
            ind += 1
            
        left_node.next = end
        start.next = cur
        return sanitel.next
        
            
            
        

File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dumb = new ListNode();
        int count = 1;
        ListNode p = dumb;
        ListNode newTail = null;
        ListNode cur = head;
        ListNode nxt;
        while(cur != null){
            nxt = cur.next;
            if(count < left){
                p.next = cur;
                cur.next = null;
                p = p.next;
            }else if(count >= left && count <= right){
                ListNode following = p.next;
                p.next = cur;
                cur.next = following;
            }else if (count == right + 1){
                p = newTail;
                p.next = cur;
                cur.next = null;
                p = p.next;
            }else{
                p.next = cur;
                cur.next = null;
                p = p.next;
            }
            if(count == left){
                newTail = cur;
            }
            cur = nxt;
            count++;
        }
        return dumb.next;
        
    }
}